### ISSUE TITLE
Stop bumping go directive unless necessitated by other dependencies

### ISSUE THREAD

**Issue by kaovilai** [chainguard-dev/apko#1454]
This repo by itself should not be enforcing minimum on other repositories importing it. Stop spreading "minimum virus"

toolchain version used will be defined outside of go.mod ideally, such as by installing a newer compatible go toolchain to ci/cd/development env.

Failing that, toolchain directive should be used instead of go directive for bumping versions to not cascade minimum versions to importing dependencies.
> [toolchain directive, in contrast to the go directive, applies only to the current module (the one defined by the go.mod file). It suggests the toolchain to be used when in that very module, and doesn't propagate to other modules.](https://github.com/golang/go/issues/70853#issuecomment-2549206026)

High profile repos that have removed/reduced minimum go patch version per user requests
- https://github.com/grpc/grpc-go/pull/7831 [motivation](https://github.com/grpc/grpc-go/pull/7624#discussion_r1757159186)

Being proactive to prevent following from reoccuring
- https://github.com/kopia/kopia/issues/4292
- https://github.com/vmware-tanzu/velero/issues/8397
- https://github.com/kubevirt/kubevirt-velero-plugin/pull/292#issuecomment-2482634796
- https://github.com/chainguard-dev/sdk/issues/147

**Comment by kaovilai** [comment]
```
@kaovilai ➜ /workspaces/apko (main) $ go mod graph | grep go@1.23
chainguard.dev/apko go@1.23.4
chainguard.dev/go-grpc-kit@v0.17.7 go@1.23.1
chainguard.dev/sdk@v0.1.29 go@1.23.3
go@1.23.4 toolchain@go1.23.4
k8s.io/apimachinery@v0.32.0 go@1.23.0
sigs.k8s.io/release-utils@v0.8.5 go@1.23
```

Nothing in this repo necessitated 1.23.1. In fact, it is shown that the lowest required by other deps is 1.23.1 as required by `chainguard.dev/sdk@v0.1.29`

**Comment by xnox** [comment]
Often there are CVEs in the standard library and the only way to enforce remediation of them in any builds is to bump the version number.

At Chainguard, we do no support building binaries with known CVEs.

**Comment by kaovilai** [comment]
The standard libraries used to build are based on toolchain setting. go directive is the wrong way to go because it affects importing modules.

https://github.com/golang/go/issues/70853#issuecomment-2549206026

Please see https://go.dev/doc/toolchain#:~:text=Go%20toolchain%20selection%C2%B6

go mod's [go directive](https://go.dev/ref/mod#go-mod-file-go) is the wrong way to go.

If your ci system lacks ability to specify the correct go version to use outside of go.mod file (my preference), use [toolchain directive](https://go.dev/ref/mod#go-mod-file-toolchain) instead please.

**Comment by kaovilai** [comment]
You do not have to build binaries with known CVEs AND force importing dependencies (those outside of chainguard) to have higher minimum versions.

Everyone including the importing modules can fix standard library CVEs through go cli updates etc WITHOUT commiting change to go.mod file.

**Comment by kaovilai** [comment]
@xnox Here are some other examples of other repos who won't "support building binaries with known CVEs." as well.

https://github.com/sigstore/rekor/pull/2323
https://github.com/google/go-github/pull/3423

A CVE might affect `go1.20.1` `go1.22.2`
and if its fixed in `1.20.2` and `1.22.3` then
the importing project should have their own autonomy whether they want to up to `1.20.2` or `1.22.3`.

The current go directive minimum bump here forces importers to use `1.22.3` rather than just using also fixed `1.20.2`

**Comment by kaovilai** [comment]
Minimum version bump should be for new language features, not CVEs imo. There are enough of us with this requirement that several repos are accepting and relaxing the minimum go version.

**Comment by imjasonh** [comment]
@kaovilai you're right, there's no practical reason we need to so aggressively bump the go version in go.mod.

However, I don't think it's that harmful to do so, and (unfortunately) we and other projects use that version to drive `actions/setup-go` to tell it which version of Go to install when building and testing ([ref](https://github.com/chainguard-dev/apko/blob/64e3ae0d1cf09e2e83dd384cd5a4f9c06d48e0a4/.github/workflows/build.yaml#L26))

Unless/until `setup-go` supports a `go-version: latest` or something that we can easily use to always get the latest version of Go, I don't see us changing our strategy for updating the go version in `go.mod`.

Thanks for the feedback.

**Comment by kaovilai** [comment]
> However, I don't think it's that harmful to do so, and (unfortunately) we and other projects use that version to drive actions/setup-go to tell it which version of Go to install when building and testing ([ref](https://github.com/chainguard-dev/apko/blob/64e3ae0d1cf09e2e83dd384cd5a4f9c06d48e0a4/.github/workflows/build.yaml#L26))

As I have suggested, at minimum, if you require version to be specified in go.mod, use toolchain directive.

> [Failing that, toolchain directive should be used instead of go directive for bumping versions to not cascade minimum versions to importing dependencies.
](https://github.com/golang/go/issues/70853#issuecomment-2549206026)

like so.
```
go 1.22

toolchain go1.22.9
```

> Unless/until setup-go supports a go-version: latest or something that we can easily use to always get the latest version of Go, I don't see us changing our strategy for updating the go version in go.mod.
```
  - uses: actions/setup-go@v5
    with:
      go-version: 'stable'
```
per https://github.com/actions/setup-go/#:~:text=%2D%20uses%3A%20actions/setup%2Dgo%40v5%0A%20%20%20%20with%3A%0A%20%20%20%20%20%20go%2Dversion%3A%20%27stable%27

**Comment by kaovilai** [comment]
@imjasonh please let me know if above info helps in solving your requirement.

**Comment by xnox** [comment]
@kaovilai 

Trying to do what you asked for

```
$ go mod tidy -go=1.17
go: chainguard.dev/sdk@v0.1.29 requires go@1.23.3, but 1.17 is requested
```

(And also set `toolchain go1.23.4`) and it doesn't work.

sdk is partially derived from private code; so will need to look into specifying "go + toolchain" there too, before this can make it here.

**Comment by kaovilai** [comment]
```
@kaovilai ➜ /workspaces/apko (main) $ go get go@1.23.3 toolchain@1.23.4 && go mod tidy && cat go.mod

module chainguard.dev/apko

go 1.23.3

toolchain go1.23.4
```
Just did this in github workspace.

**Comment by kaovilai** [comment]
fyi once you move to setup-go go-version: stable you can remove the toolchain line.

**Comment by xnox** [comment]
> ```
> @kaovilai ➜ /workspaces/apko (main) $ go get go@1.23.3 toolchain@1.23.4 && go mod tidy && cat go.mod
> 
> module chainguard.dev/apko
> 
> go 1.23.3
> 
> toolchain go1.23.4
> ```
> 
> Just did this in github workspace.

sure, but this is of limited value. given that weekly releases of sdk will continue to bump go version; and once that dep is bumped here; we would be virtually back to chasing latest go version here as well.

**Comment by kaovilai** [comment]
You would have to do the same with SDK. SDK don't need go **_directive_** bump all the time..

Once that happens, your go.mod file won't need to change until you actually bump non-**go** dependencies.

**Comment by kaovilai** [comment]
your go line will stay the same. leave the version bumping for stdlib CVEs to setup-go version: `stable`

**Comment by kaovilai** [comment]
I get it if its a dep outside of chainguard's control. So far, go mod graph results say otherwise.

I am only reaching out to repos based on deps my projects are importing directly/indirectly. Many repos already managed relax go minimum as I have highlighted in this issue.

I am advocating on behalf of other users of our projects, namely: https://github.com/kubevirt/kubevirt-velero-plugin/pull/292#issuecomment-2482634796

**Comment by kaovilai** [comment]
Per my [earlier comment](https://github.com/chainguard-dev/apko/issues/1454#issuecomment-2561735682) bumps in this repo were in fact driven by this repo itself, not other deps.

Specifically, only `chainguard.dev/apko` has dep of `go@1.23.4` from the `go mod graph grep go@1.23` output as an example.

**Comment by kaovilai** [comment]
You will eventually inherit a newer minimum from dependencies that also misunderstood go.mod go directive yes, but this repo itself should try to be good golangtizens by not adding to the issue.

**Comment by xnox** [comment]
> I am advocating on behalf of other users of our projects, namely: [kubevirt/kubevirt-velero-plugin#292 (comment)](https://github.com/kubevirt/kubevirt-velero-plugin/pull/292#issuecomment-2482634796)

i am not seeing chainguard.dev imports there, which projects are impacted that depend on chainguard.dev/apko and what are their current min go directives? where abouts are you importing apko?

**Comment by kaovilai** [comment]
That repo is the initial request that sparked my advocacy for this issue.

Regarding apko specifically, some of the projects I import have go minimum as a result of apko.

Specifically https://github.com/containers/image/issues/2664 enforced 1.22.6 from
```
containers/image
❯ go mod graph | grep go@1.22.6  
github.com/sigstore/fulcio@v1.6.4 go@1.22.6
```

So I hunt down fulcio's minimum for any repos pinning non zero go **directive** for repos contributing to this problem
https://github.com/sigstore/fulcio/tree/v1.6.4

```
@kaovilai ➜ /workspaces/fulcio (5237979) $ go mod graph | grep go@1.22
github.com/sigstore/fulcio go@1.22.6
chainguard.dev/sdk@v0.1.23 go@1.22.5 <--
github.com/letsencrypt/boulder@v0.0.0-20240620165639-de9c06129bec go@1.22.0
github.com/sigstore/sigstore@v1.8.8 go@1.22.5
github.com/sigstore/sigstore/pkg/signature/kms/aws@v1.8.8 go@1.22.5
github.com/sigstore/sigstore/pkg/signature/kms/azure@v1.8.8 go@1.22.5
github.com/sigstore/sigstore/pkg/signature/kms/gcp@v1.8.8 go@1.22.5
github.com/sigstore/sigstore/pkg/signature/kms/hashivault@v1.8.8 go@1.22.5
go@1.22.6 toolchain@go1.22.6
@kaovilai ➜ /workspaces/fulcio (5237979) $ 
```

Here we first encounter chainguard induced minimum of 1.22.5 in sdk v0.1.23

So I went and check if chainguard-dev/sdk/tree/main exhibit this non zero minimum
```
@kaovilai ➜ /workspaces/sdk (main) $ go mod graph | grep chainguard.dev/sdk | grep go@1.
chainguard.dev/sdk go@1.23.3
@kaovilai ➜ /workspaces/sdk (main) $ go mod graph | grep go@1.23
chainguard.dev/sdk go@1.23.3
chainguard.dev/apko@v0.22.2 go@1.23.2
chainguard.dev/go-grpc-kit@v0.17.7 go@1.23.1
go@1.23.3 toolchain@go1.23.3
k8s.io/apimachinery@v0.32.0 go@1.23.0
@kaovilai ➜ /workspaces/sdk (main) $ 
```
Sure enough, chainguard-dev/sdk/tree/main has a non minimum required go directive of 1.23.3 where the absolute minimum possible is 1.23.2 as required by
```
chainguard.dev/apko@v0.22.2 go@1.23.2
```

So for chainguard-dev/sdk/tree/main to drop minimum to 1.23.0 or 1.23, one needs to fix following
- `chainguard.dev/apko@v0.22.2 go@1.23.2`
- `chainguard.dev/go-grpc-kit@v0.17.7 go@1.23.1`

**Comment by kaovilai** [comment]
So TL;DR is, Any kind of SDK should try to minimize minimum go version required to build.

A library should not be forcing importers to bump versions without basis.

> [As mentioned in https://github.com/grpc/grpc-go/pull/7624#discussion_r1757159186, grpc-go being a library should not specify a patch version.](https://github.com/grpc/grpc-go/pull/7831#:~:text=As%20mentioned%20in%20this%20thread%2C%20grpc%2Dgo%20being%20a%20library%20should%20not%20specify%20a%20patch%20version.)

Your SDK have minimums coming from apko in main branch, so I am proactively notifying so you are aware that go directive should not be bumped. If you don't have non go.mod alternative to enforce a go toolchain version used, that you please [use toolchain directive](https://go.dev/ref/mod#go-mod-file-toolchain) as [intended by go authors](https://go.dev/doc/toolchain) which is meant to enforce version used to compile code, not go directive, which is a construct for language version compatibility.

**Comment by xnox** [comment]
@kaovilai we are attempting to keep go at 1.23 and bump toolchain to 1.24 in wolfictl. https://github.com/wolfi-dev/wolfictl/pull/1453/files

However doing so increases build & test requirements. As 1.24 introduces PQC TLS, and supports setting main.Version automatically, meaning we would need to do multiple builds and tests.

Also syft is one of the dependencies of wolfictl, which has already bumped to 1.24 see https://github.com/anchore/syft/commit/dce99543b25d6082cc2318c80b71e35f972c9e73#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6

wolfictl will be imported by sdk, sdk will be imported by apko, and apko will be imported by melange. Thus this will result in go 1.24.0 spreading through all our dependencies.

If you want for us to continue to support go1.23 directive please help us untagle these dependencies; convince syft to do similar separates of go & toolchain directives; or les us know which portions of sdk/modules you are after that you would want to be maintained with less deps and/or which golang toolchains.

Also in all of the above messages, concrete examples of where/how you use apko were not shown. I am closing this as not-planned for now, until there is further input or improvements/synergy to do similar splits in projects we depend on.

**Comment by xnox** [comment]
@kaovilai i think adoption of maintaining split-levels of toolchains will become wider; if you can convince golang upstream to require the two be declared separately and intentionally try to calculate the min/max versions as usable ones.
